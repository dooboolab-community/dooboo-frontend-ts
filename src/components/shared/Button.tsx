import React, { CSSProperties, ReactElement } from 'react';

import { ButtonPrimary } from '../ui/Buttons';
import styled from 'styled-components';

interface Props {
  id?: string;
  style?: CSSProperties;
  imgSrc?: string;
  text?: string;
  onClick?: () => void;
  isLoading?: boolean;
}

const ButtonWrapper = styled.div`
  flex: 1;
  height: 100%;

  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
`;

const PrimaryText = styled.span`
  font-size: 14px;
  color: ${({ theme }): string => theme.btnPrimaryFont};
  margin: auto;
`;

const LogoImg = styled.img`
  position: absolute;
  left: 20px;
  height: 20px;
  width: 20px;
  object-fit: cover;
`;

const Spinner = styled.div`
  border: 4px solid #f3f3f3; /* Light grey */
  border-top: 4px solid #3498db; /* Blue */
  border-radius: 50%;
  width: 24px;
  height: 24px;
  animation: spin 1s linear infinite;
  @keyframes spin {
    0% {
      transform: rotate(0deg);
    }
    100% {
      transform: rotate(360deg);
    }
  }
`;

function Button(props: Props): ReactElement {
  const { onClick, imgSrc, text, style, isLoading } = props;
  return (
    <ButtonPrimary style={style} onClick={onClick}>
      {isLoading ? (
        <Spinner id="spinner" />
      ) : (
        <ButtonWrapper>
          {imgSrc ? <LogoImg src={imgSrc} /> : null}
          <PrimaryText>{text}</PrimaryText>
        </ButtonWrapper>
      )}
    </ButtonPrimary>
  );
}

Button.defaultProps = {
  style: { display: 'flex', height: '60px' },
};

export default Button;
