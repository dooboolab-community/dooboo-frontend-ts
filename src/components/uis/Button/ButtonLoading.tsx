import type {ReactElement} from 'react';
import styled from '@emotion/styled';

const Spinner = styled.div`
  align-self: center;
  border: 4px solid #f3f3f3; /* Light grey */
  border-top: 4px solid #3498db; /* Blue */
  border-radius: 50%;
  width: 24px;
  height: 24px;
  animation: spin 1s linear infinite;
  @keyframes spin {
    0% {
      transform: rotate(0deg);
    }
    100% {
      transform: rotate(360deg);
    }
  }
`;

const ButtonLoading = (): ReactElement => {
  return <Spinner id="spinner" />;
};

export default ButtonLoading;
