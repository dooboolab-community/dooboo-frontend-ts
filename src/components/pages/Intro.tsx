import {ReactElement, useState} from 'react';

import Button from '../uis/Button';
import {IC_GOOGLE_W} from '../../utils/Icons';
import {User} from '../../types';
import UserCard from '../uis/UserCard';
import {device} from '../../theme';
import styled from '@emotion/styled';
import {useNavigate} from 'react-router-dom';
import {useRecoilState} from 'recoil';
import {useThemeContext} from '../../providers/ThemeProvider';
import {userRecoilState} from '../../recoil/atoms';

const Container = styled.div`
  display: flex;
  height: 100vh;
  width: 100vw;
  align-self: stretch;
  overflow: scroll;
  background: ${(props) => props.theme.background};

  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  overflow: hidden;
`;

const ButtonWrapper = styled.div`
  position: absolute;
  flex-direction: column;

  @media ${device.mobileS} {
    bottom: 40px;
    width: 85vw;
    align-self: center;
  }

  @media ${device.tablet} {
    width: 50vw;
    right: 60px;
    align-self: center;
    top: 400px;
  }
`;

function Intro(): ReactElement {
  // eslint-disable-next-line
  let timer: any;
  const navigate = useNavigate();
  const [_, setUser] = useRecoilState(userRecoilState);
  const {changeThemeType} = useThemeContext();
  const [isLoggingIn, setIsLoggingIn] = useState(false);

  const onLogin = (): void => {
    setIsLoggingIn(true);

    setUser(null);

    timer = setTimeout(() => {
      const newUser: User = {
        displayName: 'dooboolab',
        age: 30,
        job: 'developer',
      };

      setUser(newUser);
      setIsLoggingIn(false);
      clearTimeout(timer);
    }, 1000);
  };

  return (
    <Container>
      <UserCard />
      <ButtonWrapper>
        <Button
          testID="SIGN_IN"
          imgSrc={IC_GOOGLE_W}
          isLoading={isLoggingIn}
          onClick={(): void => onLogin()}
          text={<fbt desc="sign in">Sign In</fbt>}
        />
        <Button
          onClick={(): void => navigate('/temp', {})}
          text={<fbt desc="navigate">Navigate</fbt>}
        />
        <Button
          testID="CHANGE_THEME"
          onClick={(): void => changeThemeType()}
          text={<fbt desc="change theme">Change theme</fbt>}
        />
      </ButtonWrapper>
    </Container>
  );
}

export default Intro;
