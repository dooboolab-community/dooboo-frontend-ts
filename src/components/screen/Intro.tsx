import { IC_FACEBOOK_W, IC_GOOGLE_W } from '../../utils/Icons';
import React, { Component } from 'react';
import { ThemeType, device } from '../../theme';

import { AppContext } from '../../providers';
import Button from '../shared/Button';
import { Redirect } from 'react-router-dom';
import { User } from '../../types';
import { getString } from '../../../STRINGS';
import styled from 'styled-components';

const Container = styled.div`
  display: flex;
  height: 100vh;
  width: 100vw;
  align-self: stretch;
  overflow: scroll;
  background: ${(props) => props.theme.background};

  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  overflow: hidden;
`;

const ContentWrapper = styled.div`
  display: flex;
  flex-direction: column;
  @media ${device.mobileS} {
    max-width: 768px;
    height: 100vh;
    width: 100vw;
    justify-content: center;
    align-items: center;
  }

  @media ${device.tablet} {
    justify-content: flex-start;
    align-items: flex-start;
    margin-top: 400px;
  }
`;

const ButtonWrapper = styled.div`
  position: absolute;
  flex-direction: column;

  @media ${device.mobileS} {
    bottom: 40px;
    width: 85vw;
    align-self: center;
  }

  @media ${device.tablet} {
    width: 50vw;
    right: 60px;
    align-self: center;
    top: 400px;
  }
`;

const Text = styled.span`
  font-size: 18px;
  line-height: 1.5;
  font-family: sans-serif;
  color: ${(props) => props.theme.fontColor};
`;

interface Props {
  history: any;
  store?: any;
}

function Intro(props: Props) {
  let timer: any;
  const { state, dispatch } = React.useContext(AppContext);
  const [isLoggingIn, setIsLoggingIn] = React.useState(false);

  const onLogin = () => {
    dispatch({ type: 'reset-user' });
    setIsLoggingIn(true);
    timer = setTimeout(() => {
      const user: User = {
        displayName: 'dooboolab',
        age: 30,
        job: 'developer',
      };
      dispatch({ type: 'set-user', payload: user });
      setIsLoggingIn(false);
      clearTimeout(timer);
    }, 1000);
  };

  const navigate = () => {
    const location: object = {
      pathname: '/404',
      state: {},
    };
    props.history.push(location);
  };

  const changeTheme = () => {
    let payload: object;
    if (state.theme === ThemeType.LIGHT) {
      payload = {
        theme: ThemeType.DARK,
      };
    } else {
      payload = {
        theme: ThemeType.LIGHT,
      };
    }
    dispatch({
      type: 'change-theme-mode',
      payload,
    });
  };

  return (
    <Container>
      <ContentWrapper>
        <Text>{state.user.displayName}</Text>
        <Text>{state.user.age ? state.user.age : ''}</Text>
        <Text>{state.user.job}</Text>
      </ContentWrapper>
      <ButtonWrapper>
        <Button
          imgSrc={IC_GOOGLE_W}
          isLoading={isLoggingIn}
          onClick={() => onLogin()}
          text={getString('LOGIN')}
        />
        <Button
          onClick={() => navigate()}
          inverted={true}
          text={getString('NAVIGATE')}
        />
        <Button
          onClick={() => changeTheme()}
          inverted={true}
          text={getString('CHANGE_THEME')}
        />
      </ButtonWrapper>
    </Container>
  );
}

export default Intro;
