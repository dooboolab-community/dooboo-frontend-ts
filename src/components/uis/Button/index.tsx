import {CSSProperties, ReactElement} from 'react';

import ButtonLoading from './ButtonLoading';
import {ButtonPrimary} from '../Styles';
import styled from '@emotion/styled';

interface Props {
  testID?: string;
  id?: string;
  style?: CSSProperties;
  imgSrc?: string;
  text?: string | ReactElement;
  onClick?: () => void;
  isLoading?: boolean;
}

const ButtonWrapper = styled.div`
  flex: 1;
  height: 100%;

  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
`;

const PrimaryText = styled.span`
  font-size: 14px;
  color: ${({theme}) => theme.btnPrimaryFont};
  margin: auto;
`;

const LogoImg = styled.img`
  position: absolute;
  left: 20px;
  height: 20px;
  width: 20px;
  object-fit: cover;
`;

function Button(props: Props): ReactElement {
  const {testID, onClick, imgSrc, text, style, isLoading} = props;

  return (
    <ButtonPrimary data-testid={testID} style={style} onClick={onClick}>
      {isLoading ? (
        <ButtonLoading />
      ) : (
        <ButtonWrapper>
          {imgSrc ? <LogoImg src={imgSrc} /> : null}
          <PrimaryText>{text}</PrimaryText>
        </ButtonWrapper>
      )}
    </ButtonPrimary>
  );
}

Button.defaultProps = {
  style: {display: 'flex', height: '60px'},
};

export default Button;
