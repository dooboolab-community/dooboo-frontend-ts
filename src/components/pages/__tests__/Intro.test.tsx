import * as renderer from 'react-test-renderer';

import {RenderResult, cleanup, fireEvent, render} from '@testing-library/react';

import Button from '../../uis/Button';
import Intro from '../Intro';
import React from 'react';
import {createTestElement} from '../../../../test/testUtils';

const props = {};
const component = createTestElement(<Intro {...props} />);

let container;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

// test for the container page in dom
describe('[Intro] screen rendering test', () => {
  let json: any;

  it('should render outer component and snapshot matches', () => {
    json = renderer.create(component).toJSON();
    expect(json).toBeTruthy();
  });
});

describe('[Intro] Interaction', () => {
  let rendered: renderer.ReactTestRenderer;
  let root: renderer.ReactTestInstance;
  let renderResult: RenderResult;

  afterEach(cleanup);

  it('should simulate [onLogin] click with testing library', () => {
    renderResult = render(component);

    const SignInButton = renderResult.getByTestId('SIGN_IN');

    fireEvent.click(SignInButton);
  });

  it('should simulate [navigate] when clicked', () => {
    rendered = renderer.create(component);
    root = rendered.root;

    const buttons = root.findAllByType(Button);

    buttons[1].props.onClick();
    expect(1).toBeTruthy(); // TODO: Expect navigate to be called.
  });

  it('should change theme when [change theme] has been clicked', () => {
    renderResult = render(component);

    const btnChangeTheme = renderResult.getByTestId('CHANGE_THEME');
    const clickResult1 = fireEvent.click(btnChangeTheme);

    expect(clickResult1).toBe(true);

    const clickResult2 = fireEvent.click(btnChangeTheme);

    expect(clickResult2).toBe(true);
  });
});
