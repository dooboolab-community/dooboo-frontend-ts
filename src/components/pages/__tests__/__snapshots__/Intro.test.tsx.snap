// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Intro] screen rendering test should render outer component and snapshot matches 1`] = `
<div
  className="css-1nvx4il"
>
  <div
    className="css-1vx0ptc"
  >
    <span
      className="css-1l2m8ze"
    >
      
    </span>
    <span
      className="css-1l2m8ze"
    >
      
    </span>
    <span
      className="css-1l2m8ze"
    >
      
    </span>
  </div>
  <div
    className="css-1j4jnb0"
  >
    <button
      className="css-wnig0z"
      data-testid="SIGN_IN"
      onClick={[Function]}
      style={
        Object {
          "display": "flex",
          "height": "60px",
        }
      }
    >
      <div
        className="css-qfy95"
      >
        <img
          className="css-1oidkpf"
          src="test-file-stub"
        />
        <span
          className="css-6apmke"
        />
      </div>
    </button>
    <button
      className="css-wnig0z"
      onClick={[Function]}
      style={
        Object {
          "display": "flex",
          "height": "60px",
        }
      }
    >
      <div
        className="css-qfy95"
      >
        <span
          className="css-6apmke"
        />
      </div>
    </button>
    <button
      className="css-wnig0z"
      data-testid="CHANGE_THEME"
      onClick={[Function]}
      style={
        Object {
          "display": "flex",
          "height": "60px",
        }
      }
    >
      <div
        className="css-qfy95"
      >
        <span
          className="css-6apmke"
        />
      </div>
    </button>
  </div>
</div>
`;
